### ISSUE TITLE
PGM Subscriber stops receiving after a period with Wifi

### ISSUE THREAD

**Issue by drewdb** [zeromq/netmq#677]
### Environment

    NetMQ Version:    4.0.0.1
    Operating System: Windows 10
    .NET Version:   4.5  

### Expected behaviour
Application running on Wifi connection not to stop receiving after random time just like the LAN connected applications.

### Actual behaviour
The Application on Wifi receives PGM Multicast packets from a Publisher (Wifi or LAN) for a period of time then stops receiving while all LAN connected applications are still running and receiving. Sometimes the Wifi Application errors with connection forcibly closed by remote host without anyone disconnecting but usually just stops receiving without error.

### Steps to reproduce the behaviour
It seems that the wireless access point starts blocking? the Wifi applications can still publish without issues. only affects the Receive on Wifi only. WireShark the publisher is still sending, WireShark on the wifi subscriber it stops receiving without a disconnection.

**Comment by drewdb** [comment]
Not sure if anyone is viewing this anymore but it looks like if you have a PGM Publisher and a PGM Subscriber in the same app both sending and receiving, it seems to trip something in the Wireless Access Point? Not on the LAN though.

If i only use a PGM Publisher in 1 app and a PGM Subscriber in the other, the Wireless Access Point is ok.

Is anything known about this sort of behavior?

Hopefully this Library is not dead.

**Comment by drewdb** [comment]
Nothing? Anyone? Seems to be a problem from Wired to Wireless.

**Comment by somdoron** [comment]
the library is not dead, however I'm not familiar with such behavior.
I can try and help debugging it, maybe but some log statements at the correct places to understand the behavior.

**Comment by drewdb** [comment]
Hi Doron,

Busy on another project but back again. it seems that it has something to do with I Frames? If i force them it seems to be ok. so far?

Only thing i get now is Unhandled Exceptions if the app is terminated from a connection like "end process" not always but often. Can't see where this is happening? any suggestions? thanks

**Comment by drewdb** [comment]
also if the connection is terminated from somewhere other than the app in the machine (host). i can trap them in UnHandledExceptions event but it always terminates. wanted to reconnect if that happens.

any suggestions?

**Comment by drewdb** [comment]
it seems the wireless access point closes the connection when it gets flooded and invokes the unhandledexception as it would be from a different thread? if correct and these things need to be handled in the same thread? how could that be handled?

**Comment by MikaelFerland** [comment]
Hi,
I try to investigate bluescreens on my server with PGM and by doing some tests I discovered that piece of code stop working after less than five minutes without any warning or error.You can only have Wireshark to watch the thing happened.  Maybe it could help you. I hope that the thing will be fix because I really need PGM working.
```
            const int MEGABIT = 1024;
            const int MEGABYTE = 1024;
            var broadcastAddress = $"pgm://10.3.181.11;239.0.0.1:5555";
            PublisherSocket server = new PublisherSocket();

            Console.WriteLine($"[Publisher] Connecting on {broadcastAddress}...");

            server.Options.MulticastHops = 1;
            server.Options.MulticastRate = 1000 * MEGABIT;
            server.Options.MulticastRecoveryInterval = TimeSpan.FromMinutes(10);
            server.Options.SendBuffer = MEGABYTE * 10; // 10 megabyte
            server.Options.SendHighWatermark = 1000;
            server.Bind(broadcastAddress);

            var objectInBytes = new byte[50000];

            Console.WriteLine($"[Publisher] Sending data...");
            while (true)
            {
                server.SendFrame(objectInBytes);
            }
```

**Comment by drewdb** [comment]
Thanks for that. I did try WireShark but i couldn't see anything. as you said just stops without warning or error. the publisher is still sending but the receiver just stops receiving. I'm sending encoded video frames and found that if i force I Frames it seems to help but sometimes what ever that is gets triggered if the wifi adapter gets too many packets or too many large packets?

the biggest issue for me now is that if the adapter cuts the connection on 1 pc, (several wifi subscribers) they all fall over with unhandledexception. i can only think because it's not triggered within the thread?

**Comment by drewdb** [comment]
Interesting thing.

The PGM Multicast address is dropped from the ARP Table. Disposing the socket and creating a new one does nothing. only thing that works is closing both pub and sub apps and starting again.

Is NetMQ capable of doing something with the ARP Table to stop this or re-add when dropped?

**Comment by m7slabs** [comment]
@drewdb  did you solve it ?

**Comment by m7slabs** [comment]
@drewdb . I think we were working on the application of the beacons to resolve this wireless issue .
I think that there is one more thing that can be done to resolve the wireless blocking issue. using a timer with a socket

**Comment by somdoron** [comment]
If anyone can include a stackoverflow of the exception that will help.

Also take a look at the Zyre.net project, it uses beacon for discovery and communication happen on TCP channel.

**Comment by drewdb** [comment]
No not solved yet. there are no exceptions as the sub socket on the wifi just stops receiving. i can see the pub is still sending and any other subs connected via cable are still receiving and only any wireless connected subs simply stop receiving without error or exception.

Somdoron, is there an example i can explore on the Zyre beacon. i have added Zyre to show who's connected and who's left and would be great if Zyre beacon could help to sort this wireless stops receiving issue.

**Comment by drewdb** [comment]
i'm open to any suggestion m7slabs

**Comment by drewdb** [comment]
the only other thing i've noticed is that when the wireless receiver stops receiving, the pgm multicast address gets dropped from the arp table on the wireless pc. the only fix is to close the app and start again. can't seem to find an event to say that it's no longer receiving to disconnect and reconnect. tried Monitor but only shows Accepted event and nothing else. Hopefully the beacon could be a solution?

**Comment by m7slabs** [comment]
@somdoron there is no exception it just stops receiving  we are using NETMQ Poller  and Queue  but the socket seems to be stop firing the receive ready function. also the whole function is running on the separate thread. I must say it has to do something with socket. somehow if we are able to debug the raw socket data. I believe the sock just stops  receiving the data. To our wonder there is no exception its kind of stuck in somewhere @drewdb  also mentioned that it sometimes start receiving the data again 
We have checked the routers for any igmp issues but nothing. 
one workaround could be using pair parallel sockets

**Comment by m7slabs** [comment]
@drewdb  can you shoot an email to me  at m7slabs at gmail dot com. I need to discuss something with you

**Comment by m7slabs** [comment]
Just a thought
it might be something related to buffers or packet sizes

**Comment by drewdb** [comment]
m7slabs. please do not bother contacting me or posting here. i started this thread some months ago for a solution from the 
NetMQ users and somdoron on this issue.

somdoron, the TCP used by Zyre will not work for me. i used the pgm multicast to transverse vlans as my solution will have
participants on different vlan's. does the NetMQMonitor actually work with PGM Multicast?

**Comment by drewdb** [comment]
@somdoron my question most likely got lost under the last?

Are there any known issues using the Monitor with a PGM socket? I've tried "inproc://#monitor", "inproc://sub.inproc" and only seem to receive the "Accepted" event. Have subscribed to "All" events. I have noticed that the publisher seems to fire the Accepted a number of times. Not sure if this renegotiating the connection? I'm trying to see if i can receive a Disconnected or some failure event.
